File: packages/ios/Sources/Components/POIResultCard/POIResultCard.swift
import SwiftUI

/// Characters `encodeURIComponent` leaves untouched. Matching the web escaping
/// exactly keeps the identifier identical on all three platforms.
private let kozmosPOIResultIdentifierAllowed = CharacterSet(
    charactersIn: "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_.!~*'()"
)

/// Stable identifier for a POI result, shared with map markers so a pin and its
/// list row can be kept in sync from one selection ID.
///
/// Produces the same value as the web `getPOIResultDomId`.
public func kozmosPOIResultIdentifier(_ poiId: String) -> String {
    let encoded = poiId
        .addingPercentEncoding(withAllowedCharacters: kozmosPOIResultIdentifierAllowed) ?? poiId
    return "poi-result-\(encoded)"
}

/// A POI search/list result row.
///
/// Mirrors the React `POIResultCard`. Selection is reported upward only; the
/// card renders exactly the state described by `poi` and `result`.
public struct KozmosPOIResultCard: View {
    @Environment(\.kozmosAnalytics) private var trackEvent

    private let poi: KozmosPOIPresentation
    private let result: KozmosPOIResultPresentation
    private let featuredLabel: String
    private let selectionLabel: String?
    /// The floor the map shows: a result on it carries a dot before its floor.
    private let currentFloorId: String?
    private let actionsLabel: String
    private let onSelect: (String) -> Void
    private let onAction: ((KozmosPOIResultAction, String) -> Void)?

    public init(
        poi: KozmosPOIPresentation,
        result: KozmosPOIResultPresentation,
        featuredLabel: String = "Featured",
        selectionLabel: String? = nil,
        currentFloorId: String? = nil,
        actionsLabel: String = "Actions for this result",
        onSelect: @escaping (String) -> Void,
        onAction: ((KozmosPOIResultAction, String) -> Void)? = nil
    ) {
        self.poi = poi
        self.result = result
        self.featuredLabel = featuredLabel
        self.selectionLabel = selectionLabel
        self.currentFloorId = currentFloorId
        self.actionsLabel = actionsLabel
        self.onSelect = onSelect
        self.onAction = onAction
    }

    private func handleAction(_ action: KozmosPOIResultAction) {
        trackEvent(
            KozmosAnalyticsEvent(
                eventName: "poi_result_action",
                component: "POIResultCard",
                properties: [
                    "poiId": poi.id,
                    "resultIndex": result.resultIndex,
                    "action": action.rawValue
                ]
            )
        )
        onAction?(action, poi.id)
    }

    /// Shown only on the selected result: an action row on every card would
    /// be a wall of buttons, and the tap that selects is the tap that asks.
    private var visibleActions: [KozmosPOIResultActionPresentation] {
        guard result.selected, available else { return [] }
        return result.actions
    }

    var onCurrentFloor: Bool { currentFloorId != nil && result.floorId == currentFloorId }

    private var available: Bool { result.isAvailable }

    private var accessibilityDescription: String {
        if let selectionLabel { return selectionLabel }
        return [
            result.featured ? featuredLabel : nil,
            poi.name,
            poi.categoryLabel,
            poi.locationLabel,
            poi.availabilityLabel,
            result.travelEstimate?.durationLabel,
            available ? nil : result.unavailableReason
        ]
        .compactMap { $0 }
        .joined(separator: ", ")
    }

    /// An unavailable result is still readable, but must not be announced as an
    /// actionable button. `.isSelected` only applies to a selectable card.
    private var accessibilityTraits: AccessibilityTraits {
        guard available else { return [] }
        return result.selected ? [.isButton, .isSelected] : .isButton
    }

    private func handleSelect() {
        guard available else { return }
        trackEvent(
            KozmosAnalyticsEvent(
                eventName: "poi_result_selected",
                component: "POIResultCard",
                properties: [
                    "poiId": poi.id,
                    "resultIndex": result.resultIndex,
                    "featured": result.featured
                ]
            )
        )
        onSelect(poi.id)
    }

    public var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            if result.featured {
                HStack(spacing: KozmosDimensions.primitivesLayoutSpacing50) {
                    Image(systemName: "star.fill")
                        .font(KozmosTypography.caption2)
                        .accessibilityHidden(true)
                    Text(featuredLabel)
                        .font(.caption.weight(.semibold))
                }
                .foregroundColor(KozmosColors.componentsPrimaryButtonsAlertButtonForegroundContentIdle)
                .padding(.horizontal, KozmosDimensions.primitivesLayoutSpacing100)
                .padding(.vertical, KozmosDimensions.primitivesLayoutSpacing50)
                .background(KozmosColors.componentsPrimaryButtonsAlertButtonBackgroundIdle)
                .clipShape(
                    RoundedRectangle(
                        cornerRadius: KozmosDimensions.semanticsRadiusControl,
                        style: .continuous
                    )
                )
                .padding(.leading, KozmosDimensions.primitivesLayoutSpacing200)
            } else if let badge = result.badge {
                // One tab, one treatment: the prototypes draw "Popular Choice"
                // in the same amber as "Featured", so the LABEL distinguishes
                // them and the styling does not. What differs is meaning -
                // featured is the CMS's word and the map marker acts on it too.
                HStack(spacing: KozmosDimensions.primitivesLayoutSpacing50) {
                    Image(systemName: "star.fill")
                        .font(KozmosTypography.caption2)
                        .accessibilityHidden(true)
                    Text(badge.label)
                        .font(.caption.weight(.semibold))
                }
                .foregroundColor(KozmosColors.componentsPrimaryButtonsAlertButtonForegroundContentIdle)
                .padding(.horizontal, KozmosDimensions.primitivesLayoutSpacing100)
                .padding(.vertical, KozmosDimensions.primitivesLayoutSpacing50)
                .background(KozmosColors.componentsPrimaryButtonsAlertButtonBackgroundIdle)
                .clipShape(
                    RoundedRectangle(
                        cornerRadius: KozmosDimensions.semanticsRadiusControl,
                        style: .continuous
                    )
                )
                .padding(.leading, KozmosDimensions.primitivesLayoutSpacing200)
            }

            Button(action: handleSelect) {
                HStack(alignment: .center, spacing: KozmosDimensions.primitivesLayoutSpacing150) {
                    VStack(alignment: .leading, spacing: KozmosDimensions.primitivesLayoutSpacing25) {
                        Text(poi.name)
                            .font(KozmosTypography.body)
                            .foregroundColor(KozmosColors.primitivesColorsForeground100)
                            .lineLimit(1)

                        if let categoryLabel = poi.categoryLabel {
                            Text(categoryLabel)
                                .font(KozmosTypography.subheadline)
                                .foregroundColor(KozmosColors.primitivesColorsForeground500)
                                .lineLimit(1)
                        }

                        HStack(spacing: KozmosDimensions.primitivesLayoutSpacing75) {
                            // A dot before the floor when it is the one the map shows.
                            if onCurrentFloor {
                                Circle()
                                    .fill(KozmosColors.primitivesColorsTheme500)
                                    .frame(width: KozmosDimensions.primitivesLayoutSpacing75, height: KozmosDimensions.primitivesLayoutSpacing75)
                                    .accessibilityHidden(true)
                            }
                            Text(poi.locationLabel)
                                .font(KozmosTypography.subheadline)
                                .lineLimit(1)
                        }
                        .foregroundColor(KozmosColors.primitivesColorsForeground500)

                        if let availabilityLabel = poi.availabilityLabel {
                            Text(availabilityLabel)
                                .font(.caption.weight(.semibold))
                                .foregroundColor(
                                    poi.availability == .open
                                        ? KozmosColors.componentsPrimaryButtonsSuccessButtonBackgroundIdle
                                        : KozmosColors.primitivesColorsForeground500
                                )
                        }
                    }
                    .frame(maxWidth: .infinity, alignment: .leading)

                    VStack(alignment: .trailing, spacing: KozmosDimensions.primitivesLayoutSpacing100) {
                        logo

                        if let travelEstimate = result.travelEstimate {
                            Text(travelEstimate.durationLabel)
                                .font(KozmosTypography.subheadline)
                                .foregroundColor(KozmosColors.primitivesColorsForeground100)
                        }
                    }
                }
                // 80 tall: the prototype's row.
                .frame(minHeight: KozmosDimensions.primitivesLayoutSizing1000 - KozmosDimensions.primitivesLayoutSpacing150 * 2)
                .padding(.horizontal, KozmosDimensions.primitivesLayoutSpacing200)
                .padding(.vertical, KozmosDimensions.primitivesLayoutSpacing150)
                // The row has no fill of its own, so without an explicit hit
                // shape only the text and the logo are tappable and the gaps
                // between them swallow taps.
                .contentShape(Rectangle())
            }
            .buttonStyle(.plain)
            .disabled(!available)
            .opacity(available ? 1 : 0.6)
            // The collapse belongs to the SELECT ROW, not the card.
            //
            // It used to sit on the outer VStack, which flattened everything
            // inside it into one element - fine while the card was only a row,
            // and the exact SwiftUI counterpart of the nested <button> the web
            // card had: any action button added below would have been drawn on
            // screen and unreachable to VoiceOver.
            .accessibilityElement(children: .ignore)
            .accessibilityLabel(accessibilityDescription)
            .accessibilityAddTraits(accessibilityTraits)
            .accessibilityAction {
                // No-ops when unavailable; the traits above already withhold
                // the button affordance so VoiceOver does not offer the action.
                handleSelect()
            }

            if !visibleActions.isEmpty {
                Divider().overlay(KozmosColors.semanticsBorderSubtle)

                HStack(spacing: KozmosDimensions.primitivesLayoutSpacing100) {
                    ForEach(visibleActions) { entry in
                        KozmosPOIResultActionButton(entry: entry) {
                            handleAction(entry.action)
                        }
                    }
                    Spacer(minLength: 0)
                }
                .padding(.horizontal, KozmosDimensions.primitivesLayoutSpacing200)
                .padding(.vertical, KozmosDimensions.primitivesLayoutSpacing100)
                .accessibilityElement(children: .contain)
                .accessibilityLabel(actionsLabel)
            }

            if !available, let unavailableReason = result.unavailableReason {
                Divider().overlay(KozmosColors.semanticsBorderSubtle)

                Text(unavailableReason)
                    .font(KozmosTypography.caption)
                    .foregroundColor(KozmosColors.primitivesColorsForeground500)
                    .padding(.horizontal, KozmosDimensions.primitivesLayoutSpacing200)
                    .padding(.vertical, KozmosDimensions.primitivesLayoutSpacing100)
            }
        }
        .background(KozmosColors.primitivesColorsBackground0)
        .clipShape(RoundedRectangle(cornerRadius: KozmosDimensions.semanticsRadiusControl, style: .continuous))
        .overlay(
            RoundedRectangle(cornerRadius: KozmosDimensions.semanticsRadiusControl, style: .continuous)
                .stroke(
                    result.selected
                        ? KozmosColors.primitivesColorsTheme500
                        : KozmosColors.semanticsBorderSubtle,
                    lineWidth: result.selected ? 2 : 1
                )
        )
        .accessibilityIdentifier(kozmosPOIResultIdentifier(poi.id))
    }

    /// The logo when there is one, 48 at radius Control; nothing otherwise.
    @ViewBuilder
    private var logo: some View {
        if let logo = poi.logo, let url = URL(string: logo.src) {
            AsyncImage(url: url) { image in
                image.resizable().aspectRatio(contentMode: .fit)
            } placeholder: {
                KozmosColors.primitivesColorsBackground100
            }
            .frame(width: KozmosDimensions.primitivesLayoutSizing600, height: KozmosDimensions.primitivesLayoutSizing600)
            .clipShape(
                RoundedRectangle(cornerRadius: KozmosDimensions.semanticsRadiusControl, style: .continuous)
            )
            .overlay(
                RoundedRectangle(cornerRadius: KozmosDimensions.semanticsRadiusControl, style: .continuous)
                    .stroke(KozmosColors.semanticsBorderSubtle, lineWidth: 1)
            )
            .accessibilityLabel(logo.alt)
        }
    }
}

/// One action on a selected result.
///
/// Its own View rather than an inline chain: SwiftUI's type checker gave up on
/// the styling when it lived inside the card's body, and splitting the
/// expression is the fix the compiler itself asks for.
private struct KozmosPOIResultActionButton: View {
    let entry: KozmosPOIResultActionPresentation
    let action: () -> Void

    private var foreground: Color {
        entry.primary
            ? KozmosColors.componentsPrimaryButtonsThemedButtonForegroundContentIdle
            : KozmosColors.primitivesColorsForeground0
    }

    private var background: Color {
        entry.primary
            ? KozmosColors.componentsPrimaryButtonsThemedButtonBackgroundIdle
            : KozmosColors.primitivesColorsBackground0
    }

    private var border: Color {
        entry.primary ? Color.clear : KozmosColors.semanticsBorderSubtle
    }

    private var shape: RoundedRectangle {
        RoundedRectangle(
            cornerRadius: KozmosDimensions.semanticsRadiusControl,
            style: .continuous
        )
    }

    var body: some View {
        Button(entry.label, action: action)
            .buttonStyle(.plain)
            .font(.subheadline.weight(.semibold))
            .foregroundColor(foreground)
            .padding(.horizontal, KozmosDimensions.primitivesLayoutSpacing200)
            .padding(.vertical, KozmosDimensions.primitivesLayoutSpacing100)
            .background(background)
            .clipShape(shape)
            .overlay(shape.stroke(border, lineWidth: 1))
            .disabled(entry.disabled)
            .opacity(entry.disabled ? 0.6 : 1)
    }
}

File: packages/android/src/main/java/com/kozmos/components/POIResultCard/POIResultCard.kt
package com.kozmos.components.poiresultcard

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Place
import androidx.compose.material.icons.filled.Schedule
import androidx.compose.material.icons.filled.Star
import androidx.compose.material3.Divider
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.selected
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import coil.compose.AsyncImage
import com.kozmos.contracts.KozmosPOIAvailability
import com.kozmos.contracts.KozmosPOIPresentation
import com.kozmos.contracts.KozmosPOIResultAction
import com.kozmos.contracts.KozmosPOIResultActionPresentation
import com.kozmos.contracts.KozmosPOIResultPresentation
import com.kozmos.providers.KozmosAnalyticsEvent
import com.kozmos.providers.LocalKozmosAnalytics
import com.kozmos.tokens.KozmosDimensions
import com.kozmos.tokens.KozmosThemeTokens

/** Characters `encodeURIComponent` leaves untouched. */
private const val URI_COMPONENT_UNRESERVED =
    "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_.!~*'()"

/**
 * Stable identifier for a POI result, shared with map markers so a pin and its
 * list row can be kept in sync from one selection ID.
 *
 * Produces the same value as the web `getPOIResultDomId`, so an identifier
 * generated on any platform matches the other two.
 */
fun kozmosPOIResultIdentifier(poiId: String): String = buildString {
    append("poi-result-")
    for (byte in poiId.toByteArray(Charsets.UTF_8)) {
        val char = byte.toInt().toChar()
        if (byte >= 0 && URI_COMPONENT_UNRESERVED.indexOf(char) >= 0) {
            append(char)
        } else {
            append('%')
            append(((byte.toInt() and 0xFF) + 0x100).toString(16).substring(1).uppercase())
        }
    }
}

/**
 * A POI search/list result row.
 *
 * Mirrors the React `POIResultCard`. Selection is reported upward only; the
 * card renders exactly the state described by [poi] and [result].
 */
@Composable
fun KozmosPOIResultCard(
    poi: KozmosPOIPresentation,
    result: KozmosPOIResultPresentation,
    onSelect: (String) -> Unit,
    modifier: Modifier = Modifier,
    featuredLabel: String = "Featured",
    /** The floor the map shows: a result on it carries a dot before its floor. */
    currentFloorId: String? = null,
    selectionLabel: String? = null,
    actionsLabel: String = "Actions for this result",
    onAction: ((KozmosPOIResultAction, String) -> Unit)? = null
) {
    val trackEvent = LocalKozmosAnalytics.current
    val available = result.isAvailable

    // Shown only on the selected result: an action row on every card would be a
    // wall of buttons, and the tap that selects is the tap that asks.
    val visibleActions = if (result.selected && available) result.actions else emptyList()

    val accessibilityDescription = selectionLabel ?: listOfNotNull(
        poi.name,
        poi.categoryLabel,
        poi.locationLabel,
        poi.availabilityLabel,
        result.travelEstimate?.durationLabel,
        if (available) null else result.unavailableReason
    ).joinToString(", ")

    Surface(
        onClick = {
            trackEvent(
                KozmosAnalyticsEvent(
                    component = "POIResultCard",
                    eventName = "poi_result_selected",
                    properties = mapOf(
                        "poiId" to poi.id,
                        "resultIndex" to result.resultIndex.toString(),
                        "featured" to result.featured.toString()
                    )
                )
            )
            onSelect(poi.id)
        },
        modifier = modifier
            .fillMaxWidth()
            .semantics {
                contentDescription = accessibilityDescription
                selected = result.selected
            },
        enabled = available,
        shape = RoundedCornerShape(KozmosDimensions.semanticsRadiusControl),
        color = KozmosThemeTokens.primitivesColorsBackground0,
        border = BorderStroke(
            width = if (result.selected) 2.dp else 1.dp,
            color = if (result.selected) {
                KozmosThemeTokens.primitivesColorsTheme500
            } else {
                KozmosThemeTokens.semanticsBorderSubtle
            }
        )
    ) {
        Column(modifier = Modifier.fillMaxWidth()) {
            if (result.featured) {
                Row(
                    modifier = Modifier
                        .padding(
                            start = KozmosDimensions.primitivesLayoutSpacing200,
                            top = KozmosDimensions.primitivesLayoutSpacing100
                        )
                        .clip(RoundedCornerShape(KozmosDimensions.semanticsRadiusControl))
                        .background(KozmosThemeTokens.componentsPrimaryButtonsAlertButtonBackgroundIdle)
                        .padding(
                            horizontal = KozmosDimensions.primitivesLayoutSpacing100,
                            vertical = KozmosDimensions.primitivesLayoutSpacing50
                        ),
                    verticalAlignment = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.spacedBy(
                        KozmosDimensions.primitivesLayoutSpacing50
                    )
                ) {
                    Icon(
                        imageVector = Icons.Default.Star,
                        contentDescription = null,
                        tint = KozmosThemeTokens.componentsPrimaryButtonsAlertButtonForegroundContentIdle,
                        modifier = Modifier.size(14.dp)
                    )
                    Text(
                        text = featuredLabel,
                        style = MaterialTheme.typography.labelSmall,
                        fontWeight = FontWeight.SemiBold,
                        color = KozmosThemeTokens.componentsPrimaryButtonsAlertButtonForegroundContentIdle
                    )
                }
            } else if (result.badge != null) {
                // One tab, one treatment: the prototypes draw "Popular Choice"
                // in the same amber as "Featured", so the LABEL distinguishes
                // them and the styling does not. What differs is meaning --
                // featured is the CMS's word and the map marker acts on it too.
                Row(
                    modifier = Modifier
                        .padding(
                            start = KozmosDimensions.primitivesLayoutSpacing200,
                            top = KozmosDimensions.primitivesLayoutSpacing100
                        )
                        .clip(RoundedCornerShape(KozmosDimensions.semanticsRadiusControl))
                        .background(KozmosThemeTokens.componentsPrimaryButtonsAlertButtonBackgroundIdle)
                        .padding(
                            horizontal = KozmosDimensions.primitivesLayoutSpacing100,
                            vertical = KozmosDimensions.primitivesLayoutSpacing50
                        ),
                    verticalAlignment = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.spacedBy(
                        KozmosDimensions.primitivesLayoutSpacing50
                    )
                ) {
                    Icon(
                        imageVector = Icons.Default.Star,
                        contentDescription = null,
                        tint = KozmosThemeTokens.componentsPrimaryButtonsAlertButtonForegroundContentIdle,
                        modifier = Modifier.size(14.dp)
                    )
                    Text(
                        text = result.badge.label,
                        style = MaterialTheme.typography.labelSmall,
                        fontWeight = FontWeight.SemiBold,
                        color = KozmosThemeTokens.componentsPrimaryButtonsAlertButtonForegroundContentIdle
                    )
                }
            }

            // 80 tall: the prototype's row.
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .defaultMinSize(minHeight = KozmosDimensions.primitivesLayoutSizing1000)
                    .padding(horizontal = KozmosDimensions.primitivesLayoutSpacing200, vertical = KozmosDimensions.primitivesLayoutSpacing150),
                verticalAlignment = Alignment.CenterVertically,
                horizontalArrangement = Arrangement.spacedBy(
                    KozmosDimensions.primitivesLayoutSpacing150
                )
            ) {
                Column(
                    modifier = Modifier.weight(1f),
                    verticalArrangement = Arrangement.spacedBy(
                        KozmosDimensions.primitivesLayoutSpacing25
                    )
                ) {
                    Text(
                        text = poi.name,
                        style = MaterialTheme.typography.bodyLarge,
                        fontWeight = FontWeight.Normal,
                        color = KozmosThemeTokens.primitivesColorsForeground100,
                        maxLines = 1,
                        overflow = TextOverflow.Ellipsis
                    )

                    poi.categoryLabel?.let { categoryLabel ->
                        Text(
                            text = categoryLabel,
                            style = MaterialTheme.typography.bodyMedium,
                            color = KozmosThemeTokens.primitivesColorsForeground500,
                            maxLines = 1,
                            overflow = TextOverflow.Ellipsis
                        )
                    }

                    Row(
                        verticalAlignment = Alignment.CenterVertically,
                        horizontalArrangement = Arrangement.spacedBy(
                            KozmosDimensions.primitivesLayoutSpacing50
                        )
                    ) {
                        // A dot before the floor when it is the one the map shows.
                        if (currentFloorId != null && result.floorId == currentFloorId) {
                            Box(
                                modifier = Modifier
                                    .size(KozmosDimensions.primitivesLayoutSpacing75)
                                    .clip(CircleShape)
                                    .background(KozmosThemeTokens.primitivesColorsTheme500)
                            )
                        }
                        Text(
                            text = poi.locationLabel,
                            style = MaterialTheme.typography.bodyMedium,
                            color = KozmosThemeTokens.primitivesColorsForeground500,
                            maxLines = 1,
                            overflow = TextOverflow.Ellipsis
                        )
                    }

                    poi.availabilityLabel?.let { availabilityLabel ->
                        Text(
                            text = availabilityLabel,
                            style = MaterialTheme.typography.labelSmall,
                            fontWeight = FontWeight.SemiBold,
                            color = if (poi.availability == KozmosPOIAvailability.Open) {
                                KozmosThemeTokens.componentsPrimaryButtonsSuccessButtonBackgroundIdle
                            } else {
                                KozmosThemeTokens.primitivesColorsForeground500
                            }
                        )
                    }
                }

                Column(
                    horizontalAlignment = Alignment.End,
                    verticalArrangement = Arrangement.spacedBy(
                        KozmosDimensions.primitivesLayoutSpacing100
                    )
                ) {
                    POILogo(poi = poi)

                    result.travelEstimate?.let { travelEstimate ->
                        Text(
                            text = travelEstimate.durationLabel,
                            style = MaterialTheme.typography.bodyMedium,
                            color = KozmosThemeTokens.primitivesColorsForeground100
                        )
                    }
                }
            }

            if (visibleActions.isNotEmpty()) {
                Divider(color = KozmosThemeTokens.semanticsBorderSubtle)

                Row(
                    modifier = Modifier
                        .fillMaxWidth()
                        .semantics { contentDescription = actionsLabel }
                        .padding(
                            horizontal = KozmosDimensions.primitivesLayoutSpacing200,
                            vertical = KozmosDimensions.primitivesLayoutSpacing100
                        ),
                    verticalAlignment = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.spacedBy(
                        KozmosDimensions.primitivesLayoutSpacing100
                    )
                ) {
                    visibleActions.forEach { entry ->
                        KozmosPOIResultActionButton(
                            entry = entry,
                            onClick = {
                                trackEvent(
                                    KozmosAnalyticsEvent(
                                        component = "POIResultCard",
                                        eventName = "poi_result_action",
                                        properties = mapOf(
                                            "poiId" to poi.id,
                                            "resultIndex" to result.resultIndex.toString(),
                                            "action" to entry.action.value
                                        )
                                    )
                                )
                                onAction?.invoke(entry.action, poi.id)
                            }
                        )
                    }
                }
            }

            if (!available && result.unavailableReason != null) {
                Divider(color = KozmosThemeTokens.semanticsBorderSubtle)

                Text(
                    text = result.unavailableReason,
                    style = MaterialTheme.typography.bodySmall,
                    color = KozmosThemeTokens.primitivesColorsForeground500,
                    modifier = Modifier.padding(
                        horizontal = KozmosDimensions.primitivesLayoutSpacing200,
                        vertical = KozmosDimensions.primitivesLayoutSpacing100
                    )
                )
            }
        }
    }
}

/**
 * One action on a selected result.
 *
 * Its own composable so the card's body stays readable, and so the button's
 * own click target is unambiguous: the row above it is the select target, and
 * a nested clickable that shared its semantics would be unreachable to
 * TalkBack even while it drew on screen.
 */
@Composable
private fun KozmosPOIResultActionButton(
    entry: KozmosPOIResultActionPresentation,
    onClick: () -> Unit
) {
    val background = if (entry.primary) {
        KozmosThemeTokens.componentsPrimaryButtonsThemedButtonBackgroundIdle
    } else {
        KozmosThemeTokens.primitivesColorsBackground0
    }
    val foreground = if (entry.primary) {
        KozmosThemeTokens.componentsPrimaryButtonsThemedButtonForegroundContentIdle
    } else {
        KozmosThemeTokens.primitivesColorsForeground0
    }

    Surface(
        onClick = onClick,
        enabled = !entry.disabled,
        shape = RoundedCornerShape(KozmosDimensions.semanticsRadiusControl),
        color = background,
        border = if (entry.primary) {
            null
        } else {
            BorderStroke(1.dp, KozmosThemeTokens.semanticsBorderSubtle)
        }
    ) {
        Text(
            text = entry.label,
            style = MaterialTheme.typography.labelLarge,
            fontWeight = FontWeight.SemiBold,
            color = foreground,
            modifier = Modifier.padding(
                horizontal = KozmosDimensions.primitivesLayoutSpacing200,
                vertical = KozmosDimensions.primitivesLayoutSpacing100
            )
        )
    }
}

@Composable
private fun POILogo(poi: KozmosPOIPresentation) {
    // The logo when there is one, 48 at radius Control; nothing otherwise.
    val logo = poi.logo ?: return
    AsyncImage(
        model = logo.src,
        contentDescription = logo.alt,
        contentScale = ContentScale.Fit,
        modifier = Modifier
            .size(KozmosDimensions.primitivesLayoutSizing600)
            .clip(RoundedCornerShape(KozmosDimensions.semanticsRadiusControl))
    )
}
